RecoBTag/DeepFlavour/plugins/DeepFlavourTagInfoProducer.cc

```diff
@@ -0,0 +1,318 @@
+    edm::EDGetTokenT<edm::ValueMap<float>> puppi_value_map_token_;
+    edm::EDGetTokenT<edm::ValueMap<int>> pvasq_value_map_token_;
+    edm::EDGetTokenT<edm::Association<VertexCollection>> pvas_token_;
+
+    bool use_puppi_value_map_;
+    bool use_pvasq_value_map_;
+    
+};
+
+DeepFlavourTagInfoProducer::DeepFlavourTagInfoProducer(const edm::ParameterSet& iConfig) :
+  jet_radius_(iConfig.getParameter<double>("jet_radius")),
+  jet_token_(consumes<edm::View<reco::Jet> >(iConfig.getParameter<edm::InputTag>("jets"))),
+  vtx_token_(consumes<VertexCollection>(iConfig.getParameter<edm::InputTag>("vertices"))),
+  sv_token_(consumes<SVCollection>(iConfig.getParameter<edm::InputTag>("secondary_vertices"))),
+  shallow_tag_info_token_(consumes<ShallowTagInfoCollection>(iConfig.getParameter<edm::InputTag>("shallow_tag_infos"))),
+  use_puppi_value_map_(false),
+  use_pvasq_value_map_(false)
+{
+  produces<DeepFlavourTagInfoCollection>();
+
+  const auto & puppi_value_map_tag = iConfig.getParameter<edm::InputTag>("puppi_value_map");
+  if (!puppi_value_map_tag.label().empty()) {
+    puppi_value_map_token_ = consumes<edm::ValueMap<float>>(puppi_value_map_tag);
+    use_puppi_value_map_ = true;
+  }
+
+  const auto & pvas_tag = iConfig.getParameter<edm::InputTag>("vertexAssociator");
+  if (!pvas_tag.label().empty()) {
+    pvasq_value_map_token_ = consumes<edm::ValueMap<int>>(pvas_tag);
+    pvas_token_ = consumes<edm::Association<VertexCollection>>(pvas_tag);
+    use_pvasq_value_map_ = true;
+  }
+
+}
+
+
+DeepFlavourTagInfoProducer::~DeepFlavourTagInfoProducer()
+{
+}
+
+void DeepFlavourTagInfoProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions)
+{
+}
+
+void DeepFlavourTagInfoProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+
+  auto output_tag_infos = std::make_unique<DeepFlavourTagInfoCollection>();
+
+  edm::Handle<edm::View<reco::Jet>> jets;
+  iEvent.getByToken(jet_token_, jets);
+
+  edm::Handle<VertexCollection> vtxs;
+  iEvent.getByToken(vtx_token_, vtxs);
+  // reference to primary vertex
+  const auto & pv = vtxs->at(0);
+
+  edm::Handle<SVCollection> svs;
+  iEvent.getByToken(sv_token_, svs);
+
+  edm::Handle<ShallowTagInfoCollection> shallow_tag_infos;
+  iEvent.getByToken(shallow_tag_info_token_, shallow_tag_infos);
+
+  edm::Handle<edm::ValueMap<float>> puppi_value_map;
+  if (use_puppi_value_map_) { 
+    iEvent.getByToken(puppi_value_map_token_, puppi_value_map);
+  }
+
+  edm::Handle<edm::ValueMap<int>> pvasq_value_map;
+  edm::Handle<edm::Association<VertexCollection>> pvas;
+  if (use_pvasq_value_map_) { 
+    iEvent.getByToken(pvasq_value_map_token_, pvasq_value_map);
+    iEvent.getByToken(pvas_token_, pvas);
+  }
+
+  edm::ESHandle<TransientTrackBuilder> track_builder; 
+  iSetup.get<TransientTrackRecord>().get("TransientTrackBuilder", track_builder);
+
+  for (std::size_t jet_n = 0; jet_n <  jets->size(); jet_n++) {
```

Comment:
shallow_tag_infos already have knowledge of the jet: why not loop over the infos collection directly and avoid the matching logic below?

Reply:
this is what I did at the beginning but I found some problem (I do not remember in detail, I will have to look into it again) which made me use the full matching logic again

Reply:
I tried to remember harder, and the problem was that the TagInfo JetRef points to the RECO collection which is not the same as the MINIAOD/PAT collection and does not have to correspond to the same order, could it be the case?